# Venture Funding with Deep Learning

You work as a risk management associate at Alphabet Soup, a venture capital firm. Alphabet Soup’s business team receives many funding applications from startups every day. This team has asked you to help them create a model that predicts whether applicants will be successful if funded by Alphabet Soup.

The business team has given you a CSV containing more than 34,000 organizations that have received funding from Alphabet Soup over the years. With your knowledge of machine learning and neural networks, you decide to use the features in the provided dataset to create a binary classifier model that will predict whether an applicant will become a successful business. The CSV file contains a variety of information about these businesses, including whether or not they ultimately became successful.

## Instructions:

The steps for this challenge are broken out into the following sections:

* Prepare the data for use on a neural network model.

* Compile and evaluate a binary classification model using a neural network.

* Optimize the neural network model.

### Prepare the Data for Use on a Neural Network Model

Using your knowledge of Pandas and scikit-learn’s `StandardScaler()`, preprocess the dataset so that you can use it to compile and evaluate the neural network model later.

Open the starter code file, and complete the following data preparation steps:

1. Read the `applicants_data.csv` file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define your features and target variables.

2. Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.

3. Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

4. Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

> **Note** To complete this step, you will employ the Pandas `concat()` function that was introduced earlier in this course.

5. Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “IS_SUCCESSFUL”. The remaining columns should define the features dataset.

6. Split the features and target sets into training and testing datasets.

7. Use scikit-learn's `StandardScaler` to scale the features data.

### Compile and Evaluate a Binary Classification Model Using a Neural Network

Use your knowledge of TensorFlow to design a binary classification deep neural network model. This model should use the dataset’s features to predict whether an Alphabet Soup&ndash;funded startup will be successful based on the features in the dataset. Consider the number of inputs before determining the number of layers that your model will contain or the number of neurons on each layer. Then, compile and fit your model. Finally, evaluate your binary classification model to calculate the model’s loss and accuracy.

To do so, complete the following steps:

1. Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.

2. Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.

> **Hint** When fitting the model, start with a small number of epochs, such as 20, 50, or 100.

3. Evaluate the model using the test data to determine the model’s loss and accuracy.

4. Save and export your model to an HDF5 file, and name the file `AlphabetSoup.h5`.

### Optimize the Neural Network Model

Using your knowledge of TensorFlow and Keras, optimize your model to improve the model's accuracy. Even if you do not successfully achieve a better accuracy, you'll need to demonstrate at least two attempts to optimize the model. You can include these attempts in your existing notebook. Or, you can make copies of the starter notebook in the same folder, rename them, and code each model optimization in a new notebook.

> **Note** You will not lose points if your model does not achieve a high accuracy, as long as you make at least two attempts to optimize the model.

To do so, complete the following steps:

1. Define at least three new deep neural network models (the original plus 2 optimization attempts). With each, try to improve on your first model’s predictive accuracy.

> **Rewind** Recall that perfect accuracy has a value of 1, so accuracy improves as its value moves closer to 1. To optimize your model for a predictive accuracy as close to 1 as possible, you can use any or all of the following techniques:
>
> * Adjust the input data by dropping different features columns to ensure that no variables or outliers confuse the model.
>
> * Add more neurons (nodes) to a hidden layer.
>
> * Add more hidden layers.
>
> * Use different activation functions for the hidden layers.
>
> * Add to or reduce the number of epochs in the training regimen.

2. After finishing your models, display the accuracy scores achieved by each model, and compare the results.

3. Save each of your models as an HDF5 file.


In [186]:
# Imports
import pandas as pd
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder

---

## Prepare the data to be used on a neural network model

### Step 1: Read the `applicants_data.csv` file into a Pandas DataFrame. Review the DataFrame, looking for categorical variables that will need to be encoded, as well as columns that could eventually define your features and target variables.


In [187]:
# Import applicants_data.csv to Google Colab
from google.colab import files
uploaded = files.upload()

Saving applicants_data.csv to applicants_data (3).csv


In [188]:
# Read the applicants_data.csv file from the Resources folder into a Pandas DataFrame
applicant_data_df = pd.read_csv("applicants_data.csv")

# Review the DataFrame
applicant_data_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [189]:
# Review the data types associated with the columns
applicant_data_df.dtypes

EIN                        int64
NAME                      object
APPLICATION_TYPE          object
AFFILIATION               object
CLASSIFICATION            object
USE_CASE                  object
ORGANIZATION              object
STATUS                     int64
INCOME_AMT                object
SPECIAL_CONSIDERATIONS    object
ASK_AMT                    int64
IS_SUCCESSFUL              int64
dtype: object

### Step 2: Drop the “EIN” (Employer Identification Number) and “NAME” columns from the DataFrame, because they are not relevant to the binary classification model.

In [190]:
# Drop the 'EIN' and 'NAME' columns from the DataFrame
applicant_data_df = applicant_data_df.drop(columns=["EIN", "NAME"])

# Review the DataFrame
applicant_data_df.head()

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0              T10       Independent          C1000    ProductDev   
1               T3       Independent          C2000  Preservation   
2               T5  CompanySponsored          C3000    ProductDev   
3               T3  CompanySponsored          C2000  Preservation   
4               T3       Independent          C1000     Heathcare   

   ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  \
0   Association       1              0                      N     5000   
1  Co-operative       1         1-9999                      N   108590   
2   Association       1              0                      N     5000   
3         Trust       1    10000-24999                      N     6692   
4         Trust       1  100000-499999                      N   142590   

   IS_SUCCESSFUL  
0              1  
1              1  
2              0  
3              1  
4              1

### Step 3: Encode the dataset’s categorical variables using `OneHotEncoder`, and then place the encoded variables into a new DataFrame.

In [191]:
# Create a list of categorical variables
categorical_variables = ["APPLICATION_TYPE", "AFFILIATION", "CLASSIFICATION", "USE_CASE", "ORGANIZATION", "STATUS", "INCOME_AMT", "SPECIAL_CONSIDERATIONS"]

# Display the categorical variables list
categorical_variables

['APPLICATION_TYPE',
 'AFFILIATION',
 'CLASSIFICATION',
 'USE_CASE',
 'ORGANIZATION',
 'STATUS',
 'INCOME_AMT',
 'SPECIAL_CONSIDERATIONS']

In [192]:
# Create a OneHotEncoder instance
enc = OneHotEncoder(sparse=False)

In [193]:
# Encode the categorcal variables using OneHotEncoder
encoded_data = enc.fit_transform(applicant_data_df[categorical_variables])

/usr/local/lib/python3.10/dist-packages/sklearn/preprocessing/_encoders.py:868: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


In [194]:
pip install --upgrade scikit-learn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [195]:
# Create a DataFrame with the encoded variables
encoded_df = pd.DataFrame(
    encoded_data,
    columns=enc.get_feature_names_out(categorical_variables)
)

# Review the DataFrame
encoded_df.head()

APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  APPLICATION_TYPE_T13  \
0                   1.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  APPLICATION_TYPE_T17  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  APPLICATION_TYPE_T25  \
0                   0.0                  0.0                   0.0   
1                   0.0                  0.0                   0.0   
2                   0.0                  0.0                   0.0   
3                   0.0                  0.0                   0.0   
4                   0.0                  0.0                   0.0   

   APPLICATION_TYPE_T29  ...  INCOME_AMT_1-9999  INCOME_AMT_10000-24999  \
0                   0.0  ...                0.0                     0.0   
1                   0.0  ...                1.0                     0.0   
2                   0.0  ...                0.0                     0.0   
3                   0.0  ...                0.0                     1.0   
4                   0.0  ...                0.0                     0.0   

   INCOME_AMT_100000-499999  INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  \
0                       0.0                 0.0               0.0   
1                       0.0                 0.0               0.0   
2                       0.0                 0.0               0.0   
3                       0.0                 0.0               0.0   
4                       1.0                 0.0               0.0   

   INCOME_AMT_25000-99999  INCOME_AMT_50M+  INCOME_AMT_5M-10M  \
0                     0.0              0.0                0.0   
1                     0.0              0.0                0.0   
2                     0.0              0.0                0.0   
3                     0.0              0.0                0.0   
4                     0.0              0.0                0.0   

   SPECIAL_CONSIDERATIONS_N  SPECIAL_CONSIDERATIONS_Y  
0                       1.0                       0.0  
1                       1.0                       0.0  
2                       1.0                       0.0  
3                       1.0                       0.0  
4                       1.0                       0.0  

[5 rows x 116 columns]

### Step 4: Add the original DataFrame’s numerical variables to the DataFrame containing the encoded variables.

> **Note** To complete this step, you will employ the Pandas `concat()` function that was introduced earlier in this course.



This code is concatenating two data frames - `applicant_data_df` and `encoded_df` - along the columns axis (axis=1). The `drop` method is used to remove the columns specified in the `categorical_variables` list from `applicant_data_df`. The resulting data frame is then concatenated with `encoded_df`.

`encoded_df` likely contains the categorical variables from `applicant_data_df` that have been encoded using a technique such as one-hot encoding or label encoding.

`pd.concat` is a function from the pandas library that allows for the concatenation of two or more data frames. The `axis` parameter specifies the axis along which the data frames should be concatenated. The default value for `axis` is 0, which concatenates the data frames along the rows axis.

In [196]:
# Add the numerical variables from the original DataFrame to the one-hot encoding DataFrame
encoded_df = pd.concat([
    applicant_data_df.drop(columns=categorical_variables),
    encoded_df
], axis=1)

# Review the DataFrame
encoded_df.head()

ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_T10  APPLICATION_TYPE_T12  \
0     5000              1                   1.0                   0.0   
1   108590              1                   0.0                   0.0   
2     5000              0                   0.0                   0.0   
3     6692              1                   0.0                   0.0   
4   142590              1                   0.0                   0.0   

   APPLICATION_TYPE_T13  APPLICATION_TYPE_T14  APPLICATION_TYPE_T15  \
0                   0.0                   0.0                   0.0   
1                   0.0                   0.0                   0.0   
2                   0.0                   0.0                   0.0   
3                   0.0                   0.0                   0.0   
4                   0.0                   0.0                   0.0   

   APPLICATION_TYPE_T17  APPLICATION_TYPE_T19  APPLICATION_TYPE_T2  ...  \
0                   0.0                   0.0                  0.0  ...   
1                   0.0                   0.0                  0.0  ...   
2                   0.0                   0.0                  0.0  ...   
3                   0.0                   0.0                  0.0  ...   
4                   0.0                   0.0                  0.0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                0.0                     0.0                       0.0   
1                1.0                     0.0                       0.0   
2                0.0                     0.0                       0.0   
3                0.0                     1.0                       0.0   
4                0.0                     0.0                       1.0   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                 0.0               0.0                     0.0   
1                 0.0               0.0                     0.0   
2                 0.0               0.0                     0.0   
3                 0.0               0.0                     0.0   
4                 0.0               0.0                     0.0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0              0.0                0.0                       1.0   
1              0.0                0.0                       1.0   
2              0.0                0.0                       1.0   
3              0.0                0.0                       1.0   
4              0.0                0.0                       1.0   

   SPECIAL_CONSIDERATIONS_Y  
0                       0.0  
1                       0.0  
2                       0.0  
3                       0.0  
4                       0.0  

[5 rows x 118 columns]

### Step 5: Using the preprocessed data, create the features (`X`) and target (`y`) datasets. The target dataset should be defined by the preprocessed DataFrame column “IS_SUCCESSFUL”. The remaining columns should define the features dataset.



In [197]:
# Define the target set y using the IS_SUCCESSFUL column
y = encoded_df["IS_SUCCESSFUL"].values

# Display a sample of y
y[:5]

array([1, 1, 0, 1, 1])

In [198]:
# Define features set X by selecting all columns but IS_SUCCESSFUL
X = encoded_df.drop(columns=["IS_SUCCESSFUL"]).values

# Review the features DataFrame
X[:5]

array([[5.0000e+03, 1.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        1.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        1.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00,
        0.0000e+00, 0.0000e+00, 0.0000e+00, 0.00

### Step 6: Split the features and target sets into training and testing datasets.


In [199]:
# Split the preprocessed data into a training and testing dataset
# Assign the function a random_state equal to 1
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

### Step 7: Use scikit-learn's `StandardScaler` to scale the features data.

In [200]:
# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler to the features training dataset
X_scaler = scaler.fit(X_train)

# Fit the scaler to the features training dataset
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

---

## Compile and Evaluate a Binary Classification Model Using a Neural Network

### Step 1: Create a deep neural network by assigning the number of input features, the number of layers, and the number of neurons on each layer using Tensorflow’s Keras.

> **Hint** You can start with a two-layer deep neural network model that uses the `relu` activation function for both layers.


In [201]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train[0])

# Review the number of features
number_input_features

117

This code sets the number_output_neurons variable to 1, which is the number of neurons in the output layer of the neural network. This is appropriate for a binary classification problem, where the model is predicting between two options.

In [202]:
# Define the number of neurons in the output layer
number_output_neurons = 1

In [203]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1 = (number_input_features + number_output_neurons) // 2

# Review the number hidden nodes in the first layer
hidden_nodes_layer1

59

In [204]:
nn = Sequential()

# Add the first hidden layer
nn.add(Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation='relu'))
# Review the number hidden nodes in the first layer
hidden_nodes_layer1

59

In [205]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2 = (hidden_nodes_layer1 + number_output_neurons) // 2

# Review the number hidden nodes in the second layer
hidden_nodes_layer2

30

In [206]:
nn = Sequential()
# Add the second hidden layer
nn.add(Dense(units=hidden_nodes_layer2, activation='relu'))
# Review the number hidden nodes in the second layer
hidden_nodes_layer2

30

In [207]:
# Create the Sequential model instance
nn = tf.keras.models.Sequential()

In [208]:
# Add the first hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation='relu'))

In [209]:
# Add the second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation='relu'))

In [210]:
# Add the output layer to the model specifying the number of output neurons and activation function
nn.add(tf.keras.layers.Dense(units=number_output_neurons, activation='sigmoid'))

In [211]:
# Display the Sequential model summary
nn.summary()

Model: "sequential_16"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_33 (Dense)            (None, 59)                6962      
                                                                 
 dense_34 (Dense)            (None, 30)                1800      
                                                                 
 dense_35 (Dense)            (None, 1)                 31        
                                                                 
Total params: 8,793
Trainable params: 8,793
Non-trainable params: 0
_________________________________________________________________


Sequential model has three layers:

The first layer (dense_6) is a Dense layer with 59 neurons. It has 6962 parameters (weights and biases). The output shape is (None, 59), where None is a placeholder for the batch size, and 59 is the number of neurons.

The second layer (dense_7) is another Dense layer with 30 neurons. It has 1800 parameters. The output shape is (None, 30).

The third layer (dense_8) is the output layer of the model. It has 1 neuron (as it's common for binary classification problems) and 31 parameters. The output shape is (None, 1).

In total, the model has 8793 trainable parameters. These are the parameters that the model will learn from the data during training.

### Step 2: Compile and fit the model using the `binary_crossentropy` loss function, the `adam` optimizer, and the `accuracy` evaluation metric.


In [212]:
# Compile the Sequential model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

- loss='binary_crossentropy' sets the loss function to binary cross entropy, which is suitable for binary classification problems.
- optimizer='adam' sets the optimizer to Adam, which is a popular optimizer for many types of neural networks.
- metrics=['accuracy'] tells the model to evaluate its performance using accuracy as the metric.

In [213]:
# Fit the model
model_history = nn.fit(X_train, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 3s 3ms/step - loss: 43050.3750 - accuracy: 0.5085
Epoch 2/50
804/804 [==============================] - 3s 3ms/step - loss: 7162.6943 - accuracy: 0.4909
Epoch 3/50
804/804 [==============================] - 4s 5ms/step - loss: 44991.1484 - accuracy: 0.5147
Epoch 4/50
804/804 [==============================] - 3s 4ms/step - loss: 44373.6367 - accuracy: 0.5014
Epoch 5/50
804/804 [==============================] - 3s 4ms/step - loss: 19534.7168 - accuracy: 0.5203
Epoch 6/50
804/804 [==============================] - 4s 5ms/step - loss: 36025.0273 - accuracy: 0.5063
Epoch 7/50
804/804 [==============================] - 5s 6ms/step - loss: 9344.4805 - accuracy: 0.5471
Epoch 8/50
804/804 [==============================] - 3s 4ms/step - loss: 18278.9668 - accuracy: 0.5117
Epoch 9/50
804/804 [==============================] - 4s 5ms/step - loss: 6198.2373 - accuracy: 0.5219
Epoch 10/50
804/804 [==============================] - 5s 6ms/step 

- X_train and y_train are the training data and labels respectively.
- epochs=50 means that the training process will go through the entire dataset 50 times. Each pass through the entire dataset is known as an epoch.

### Step 3: Evaluate the model using the test data to determine the model’s loss and accuracy.


In [214]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test, y_test, verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - loss: 0.6915 - accuracy: 0.5292 - 461ms/epoch - 2ms/step
Loss: 0.691455602645874, Accuracy: 0.5292128324508667


### Step 4: Save and export your model to an HDF5 file, and name the file `AlphabetSoup.h5`.


In [215]:
# Set the model's file path
file_path = "AlphabetSoup.h5"

# Export your model to a HDF5 file
nn.save(file_path)

# Download your model to your computer
files.download(file_path)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

---

## Optimize the neural network model


### Step 1: Define at least three new deep neural network models (resulting in the original plus 3 optimization attempts). With each, try to improve on your first model’s predictive accuracy.

> **Rewind** Recall that perfect accuracy has a value of 1, so accuracy improves as its value moves closer to 1. To optimize your model for a predictive accuracy as close to 1 as possible, you can use any or all of the following techniques:
>
> * Adjust the input data by dropping different features columns to ensure that no variables or outliers confuse the model.
>
> * Add more neurons (nodes) to a hidden layer.
>
> * Add more hidden layers.
>
> * Use different activation functions for the hidden layers.
>
> * Add to or reduce the number of epochs in the training regimen.


### Alternative Model 1

In [216]:
# Define the the number of inputs (features) to the model
number_input_features = X_train.shape[1]

# Review the number of features
print(number_input_features)

117


In [217]:
# create a numpy array with 117 columns
X_train_array = np.random.rand(50,117)

# convert numpy array to pandas dataframe
X_train_df = pd.DataFrame(X_train_array)

# Define the the number of inputs (features) to the model
number_input_features = len(X_train_df.columns)

# Review the number of features
number_input_features


117

In [218]:
# Define the the number of inputs (features) to the model
number_input_features = X_train_df.shape[1]

# Review the number of features
print(number_input_features)


117


In [219]:
# Define the number of neurons in the output layer
number_output_neurons_A1 = 1

For binary classification problems, you typically have one neuron in the output layer. This neuron would output a value between 0 and 1, representing the probability of the positive class.

For multi-class classification problems, you would typically have as many neurons as there are classes. Each neuron would output the probability of a specific class.

For regression problems, you would typically have one neuron in the output layer. This neuron would output the predicted value.

In [220]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A1 =  int(number_input_features / 2)

# Review the number of hidden nodes in the first layer
print(hidden_nodes_layer1_A1)

58


In [221]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2_A1 =  int(hidden_nodes_layer1_A1 / 2)

# Review the number of hidden nodes in the second layer
print(hidden_nodes_layer2_A1)

29


In [222]:
# Define the number of hidden nodes for the third hidden layer
hidden_nodes_layer3_A1 =  int(hidden_nodes_layer2_A1 / 2)

# Review the number of hidden nodes in the third layer
print(hidden_nodes_layer3_A1)

14


In [223]:
# Create the Sequential model instance
nn_A1 = tf.keras.models.Sequential()

In [224]:
# Create the Sequential model instance
nn_A1 = tf.keras.models.Sequential()

# First hidden layer
nn_A1.add(tf.keras.layers.Dense(units=hidden_nodes_layer1_A1, input_dim=number_input_features, activation="relu"))

# Second hidden layer
nn_A1.add(tf.keras.layers.Dense(units=hidden_nodes_layer2_A1, activation="relu"))

# Third hidden layer
nn_A1.add(tf.keras.layers.Dense(units=hidden_nodes_layer3_A1, activation="relu"))

# Output layer
nn_A1.add(tf.keras.layers.Dense(units=number_output_neurons_A1, activation="sigmoid"))

# Check the structure of the model
nn_A1.summary()

Model: "sequential_18"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_36 (Dense)            (None, 58)                6844      
                                                                 
 dense_37 (Dense)            (None, 29)                1711      
                                                                 
 dense_38 (Dense)            (None, 14)                420       
                                                                 
 dense_39 (Dense)            (None, 1)                 15        
                                                                 
Total params: 8,990
Trainable params: 8,990
Non-trainable params: 0
_________________________________________________________________


In [225]:
# Compile the Sequential model
nn_A1.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])


In [226]:
# Fit the model using 50 epochs and the training data
fit_model_A1 = nn_A1.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 3s 2ms/step - loss: 0.5732 - accuracy: 0.7196
Epoch 2/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5512 - accuracy: 0.7300
Epoch 3/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5487 - accuracy: 0.7309
Epoch 4/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5471 - accuracy: 0.7308
Epoch 5/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5450 - accuracy: 0.7324
Epoch 6/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5441 - accuracy: 0.7350
Epoch 7/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5434 - accuracy: 0.7335
Epoch 8/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5422 - accuracy: 0.7354
Epoch 9/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5415 - accuracy: 0.7364
Epoch 10/50
804/804 [==============================] - 2s 3ms/step - loss: 0.5407 - accuracy: 0.7357

#### Alternative Model 2

In [232]:
# Define the the number of inputs (features) to the model
number_input_features = len(X_train[0])

# Review the number of features
number_input_features

117

In [233]:
# Define the number of neurons in the output layer
number_output_neurons_A2 = 1

In [234]:
# Define the number of hidden nodes for the first hidden layer
hidden_nodes_layer1_A2 =  (number_input_features + number_output_neurons_A2) // 2

# Review the number of hidden nodes in the first layer
hidden_nodes_layer1_A2


59

In [235]:
# Define the number of hidden nodes for the second hidden layer
hidden_nodes_layer2_A2 =  (hidden_nodes_layer1_A2 + number_output_neurons_A2) // 2

# Review the number of hidden nodes in the second layer
hidden_nodes_layer2_A2

30

In [237]:
# Create the Sequential model instance
nn_A2 = Sequential()

In [238]:
# First hidden layer
nn_A2.add(Dense(units=hidden_nodes_layer1_A2, input_dim=number_input_features, activation="relu"))

# Output layer
nn_A2.add(Dense(units=number_output_neurons_A2, activation="sigmoid"))

# Check the structure of the model
nn_A2.summary()

Model: "sequential_20"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_40 (Dense)            (None, 59)                6962      
                                                                 
 dense_41 (Dense)            (None, 1)                 60        
                                                                 
Total params: 7,022
Trainable params: 7,022
Non-trainable params: 0
_________________________________________________________________


In [241]:
# Compile the Sequential model
nn_A2.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [242]:
# Fit the model
fit_model_A2 = nn_A2.fit(X_train_scaled, y_train, epochs=50)

Epoch 1/50
804/804 [==============================] - 4s 2ms/step - loss: 0.5958 - accuracy: 0.7142
Epoch 2/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5590 - accuracy: 0.7286
Epoch 3/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5544 - accuracy: 0.7295
Epoch 4/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5522 - accuracy: 0.7306
Epoch 5/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5499 - accuracy: 0.7301
Epoch 6/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5498 - accuracy: 0.7315
Epoch 7/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5486 - accuracy: 0.7313
Epoch 8/50
804/804 [==============================] - 2s 2ms/step - loss: 0.5481 - accuracy: 0.7311
Epoch 9/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5467 - accuracy: 0.7330
Epoch 10/50
804/804 [==============================] - 1s 2ms/step - loss: 0.5469 - accuracy: 0.7328

### Step 2: After finishing your models, display the accuracy scores achieved by each model, and compare the results.

In [243]:
print("Original Model Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled, y_test, verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Original Model Results
268/268 - 0s - loss: 1.2861 - accuracy: 0.5277 - 365ms/epoch - 1ms/step
Loss: 1.2860727310180664, Accuracy: 0.5276967883110046


In [244]:
print("Alternative Model 1 Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn_A1.evaluate(X_test_scaled, y_test, verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Alternative Model 1 Results
268/268 - 0s - loss: 0.5529 - accuracy: 0.7300 - 449ms/epoch - 2ms/step
Loss: 0.5529299974441528, Accuracy: 0.7300291657447815


In [245]:
print("Alternative Model 2 Results")

# Evaluate the model loss and accuracy metrics using the evaluate method and the test data
model_loss, model_accuracy = nn_A2.evaluate(X_test_scaled, y_test, verbose=2)

# Display the model loss and accuracy results
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

Alternative Model 2 Results
268/268 - 1s - loss: 0.5577 - accuracy: 0.7293 - 645ms/epoch - 2ms/step
Loss: 0.5576784014701843, Accuracy: 0.7293294668197632


### Step 3: Save each of your alternative models as an HDF5 file.


In [246]:
# Set the file path for the first alternative model
file_path = "AlphabetSoup_A1.h5"

# Export your model to a HDF5 file
nn_A1.save(file_path)

# Download the model to your computer
files.download(file_path)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [247]:
# Set the file path for the second alternative model
file_path = "AlphabetSoup_A2.h5"

# Export your model to a HDF5 file
nn_A2.save(file_path)

# Download the model to your computer
files.download(file_path)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>